### DISCUSSION TITLE
Fixup method name in decompiler & function window

### DISCUSSION THREAD

**Discussion by commonuserlol** [NationalSecurityAgency/ghidra#6761] [General]
Both name and signature defined by user's script. For example, demangled name is `Core::Main::Init()` and signature is `void Main_Init()`. Because there's can be more than one `Init`, signature contains class name (`Main`).

Name set using `DemanglerCmd`, `createLabel` as fallback and signature with `ApplyFunctionSignatureCmd`.

In **decompiler** it looks like `Core::Main::Main_Init`, instead of expected `Core::Main::Init`.

In **functions window** ghidra shows signature (`Main_Init`) in `Name` field (howewer, if I manually remove this column and select `Function Name` below, it will show demangled name as in decompiler).

So, my question is can ghidra automatically detect and remove duplicates in name, like `Core::Main::Main_Init` -> `Core::Main::Init`. If no, is there script example to do it?

**Comment by dragonmacher** [comment]
It sounds like you are running into an issue where you have set the label at the function's entry point to be the demangled name (`Core::Main::Init()`) and you have modified the signature to be something more specific (`void Main_Init()`).

There is no feature in Ghidra that will attempt to consolidate multiple labels at an address with each other or with the function signature.

If you have the same pattern repeated in your program, then you should be able to write a script to detect this condition and make any fixes you need, for example, to change the function signature to match the demangled name, if that is preferred.

**Comment by commonuserlol** [comment]
> for example, to change the function signature to match the demangled name, if that is preferred.

Well, this signature used in `Main_Class` (`class Main` in source code) as class member. If there's `Core::Server::Init` and I change both signatures to `void Init()`, I'll get a wrong signature. IDA ignores signature (probably) and uses name in decompiler, can Ghidra do it?

**Comment by dragonmacher** [comment]
> can Ghidra do it?

No way I know of.  Ghidra currently enforces, at the function's entry point, that the primary label matches the function signature's name.

**Reply by commonuserlol** [reply]
At entry point only? Or in all functions?

**Reply by dragonmacher** [reply]
Each function.   The first address for each function is what we call the function's entry point (not to be confused with the program's entry point).     For each function, you may have multiple labels anywhere in the function body, including the function's entry point.    

The special case for the function's entry point is that the *primary* label will have the same name as the function signature.  If you change the signature's name or the primary label's name, they will both be updated so they match.

**Reply by commonuserlol** [reply]
Oh, so currently no way to deal with doubled class name in decompiler right?

**Reply by dragonmacher** [reply]
If I understand your issue correctly, then we currently have no way of doing what you are asking.

**Reply by commonuserlol** [reply]
Ehh I see, thanks
